RavenDB-22307 Corax enhancements

maciejaszyk opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-22307

### Additional description

- AndNot memory usage optimization and span fix (we were creating an invalid span)
- Remove UnaryMatch
- Fix MultiUnaryMatch
- Pass Slices/FieldMetadata as constant reference instead of copying

### Type of change

- [x] Bug fix
- [ ] Regression bug fix
- [x] Optimization
- [ ] New feature

### How risky is the change?

- [ ] Low 
- [x] Moderate 
- [ ] High
- [ ] Not relevant

### Backward compatibility

- [x] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [ ] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [x] No documentation update is needed 

### Testing by Contributor

- [x] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [ ] It has been verified by manual testing

### Testing by RavenDB QA team

- [ ] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work

- [ ] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [x] No UI work is needed

maciejaszyk committed:
RavenDB-22307 `AndNot `- memory usage optimization.

maciejaszyk committed:
RavenDB-22307 Explicitly block ptr arithmetic overflow.

maciejaszyk committed:
RavenDB-22307 Replace UnaryMatch with MultiUnaryMatch to avoid code-duplication.

maciejaszyk committed:
RavenDB-22307 Pass Slices/FieldMetadata as readonly reference instread of copying whole struct.

maciejaszyk committed:
RavenDB-22307 Avoid allocations in Fill method. Fix of AndWith implementation.

maciejaszyk committed:
RavenDB-22307 GrowableBuffer Tests & Invalid pointer when buffer is growing.

maciejaszyk committed:
RavenDB-22307 Since Slice is not readonly let not pass it with `in` to avoid copying on access

maciejaszyk committed:
RavenDB-22307 Use SliceComparer for Suggestions handling.

maciejaszyk committed:
RavenDB-22307 Call `Fill` from MultiUnaryMatch instead `_inner.Fill()` since we've to perform filtering on inner results.

arekpalinski commented:
test this in debug please

redknightlois commented on src/Corax/Querying/Matches/MultiUnaryMatch.cs:
```diff
@@ -382,115 +394,145 @@ public static bool Compare(ReadOnlySpan<byte> sx, ReadOnlySpan<byte> sy)
+                            MultiUnaryItem.DataType.Double => comparer.CompareNumerical(reader),
+                            _ => ThrowUnknownComparer(comparer)
+                        };
                         
```
Can't we know this before the loop and therefore use a managed function pointer to execute the correct one instead? I think the best bet would be a proxy struct with a templated function call which will ensure the entire while get's executed with specialized and packed code.

maciejaszyk replied:
I'm not sure what you have in mind here. We could build a struct, but overall the call would look like `struct.Compare(reader, comparerIdx)`. Do you think it could increase the perf?

ayende replied:
I think he meant something like this:

```
delegate <in EntryTermsReader it, bool> compareDelegate = comparer.Type switch
{
    MultiUnaryItem.DataType.Slice => comparer.CompareLiteral,
    MultiUnaryItem.DataType.Long => comparer.CompareNumerical,
    MultiUnaryItem.DataType.Double => comparer.CompareNumerical,
    _ => ThrowUnknownComparer(comparer)
};
while (reader.FindNext(currentFieldRootPage))
{
    var cmpResult = ccompareDelegate(reader);
```

redknightlois commented on src/Corax/Utils/GrowableBuffer.cs:
```diff
@@ -0,0 +1,89 @@
+        return size - (size % sizeof(long));
+    }
+}
+
```
Maybe we should also add a conservative policy. Starts with just a few elements (16) and grow from there.

redknightlois commented on src/Corax/Utils/GrowableBuffer.cs:
```diff
@@ -0,0 +1,89 @@
+}
+
+public readonly struct Slowly : IBufferGrowth
+{
```
I don't really like the name `Slowly` because it is not really slow. It actually starts with a high 16Mb baseline and grow progressively. Maybe `Progressive`, `ThresholdGrowth` or something on that line would work. 

There are a few other interesting policies that we have seen and probably we should do the work to start changing that in the rest of the codebase:

LinearGrowth: This policy could increase the buffer size by a fixed amount each time the threshold is exceeded. This might be appropriate when the growth is expected to be steady and predictable.

ExponentialGrowth: This policy could start with a smaller initial size and then grow exponentially, doubling the size each time the threshold is exceeded. This might be useful for scenarios where the expected growth is rapid and unpredictable.

BoundedGrowth<T>: This policy would put a bound to the size and could be built on top of the other strategies.

maciejaszyk replied:
The name was terrible :) Thanks for suggestions

redknightlois commented on src/Corax/Utils/GrowableBuffer.cs:
```diff
@@ -0,0 +1,89 @@
+﻿using System;
+using Sparrow.Server;
+namespace Corax.Utils;
+
```
Since we know that lots of bugs happen because of growable buffers, this should go as internal in `Sparrow` or `Sparrow.Server` if we know for a fact that Raven.Client does not use growable buffers.

maciejaszyk replied:
Yes, I moved it into `Sparrow.Server/Utils`

redknightlois reviewed (changes requested):
Overall great work, as I have been following the development nothing too new here. There are a few things I would still adjust. Considering the effort we are doing to standarize Voron it would make sense to apply the same criteria for Corax.

maciejaszyk committed:
RavenDB-22307
- Moving `GrowableBuffer` to `Sparrow.Server`
- Renaming method in `GrowableBuffer`
- Applying `GrowableBuffer` to `AndNot` primitive.

redknightlois commented:
LGTM
